### PR TITLE
Improve URL detection

### PR DESCRIPTION
**Pull request by arza-zara**
Fixed:
- naive URL detection for some cases (#2132, related #5038). The [rules for allowed characters](https://www.iana.org/assignments/idna-tables-11.0.0/idna-tables-11.0.0.xhtml) in internationalized domains are complicated, I check for simple cases with regex.
- scheme detction with %20 (#2429)
- searching e-mail addresses with search engine
- searching for search engine name when `url.open_base_url=true` (#5072)
- `url.open_base_url` with `url.auto_search=never`
- crash when search engine has invalid URL (#970)

### CODE REVIEW

**Comment by The-Compiler** [comment]
Could you do another merge from `master` please? I messed up some stuff there, but it should all be fixed now.

File: tests/unit/utils/test_urlutils.py [hunk 1]
```diff
@@ -416,6 +424,18 @@ def test_is_url(config_stub, fake_dns,
             auto_search))
 
 
+@pytest.mark.parametrize('auto_search, open_base_url, is_url', [
+    ('naive', True, False),
+    ('naive', False, False),
+    ('never', True, False),
```
Comment (arza-zara) [inline tests/unit/utils/test_urlutils.py]: I fixed that kind of accidentially so I added a test. There's still a dns lookup with `url.auto_search=dns` though, but maybe that's enough for now.

File: qutebrowser/utils/urlutils.py [hunk 2]
```diff
@@ -148,7 +150,9 @@ def _is_url_naive(urlstr: str) -> bool:
         return False
 
     host = url.host()
-    return '.' in host and not host.endswith('.')
+    tld = r'\.([^.0-9_-]+|xn--[a-z0-9-]+)$'
+    forbidden = r'[\u0000-\u002c\u002f-\u002f\u003a-\u0060\u007b-\u00b6]'
```
Comment (The-Compiler) [inline qutebrowser/utils/urlutils.py]: This probably should be `\u002e-\u002f` instead of `\u002f-\u002f`, no?

**Comment by The-Compiler** [comment]
Oh, I somewhat accidentally merged this already. :sweat_smile: 

I'm assuming you did indeed mean `u002e` in the regex above, so I pushed a commit changing that.

**Comment by arza-zara** [comment]
No, `\u002e` is dot so I think it should be allowed :wink:. It doesn't have to be in a range though.

**Comment by The-Compiler** [comment]
Oh, damn :laughing: Fixed! I was just looking at "IDNA Derived Properties" in the [list you linked](https://www.iana.org/assignments/idna-tables-11.0.0/idna-tables-11.0.0.xhtml).

File: qutebrowser/utils/urlutils.py [hunk 3]
```diff
@@ -277,10 +276,10 @@ def is_url(urlstr):
             return engine is None
 
     if not qurl_userinput.isValid():
-        # This will also catch URLs containing spaces.
+        # This will also catch non-URLs containing spaces.
         return False
 
-    if _has_explicit_scheme(qurl):
+    if _has_explicit_scheme(qurl) and ' ' not in urlstr:
```
[the remark this answers is not in the document]
Comment (arza-zara) [reply qutebrowser/utils/urlutils.py]: It's kept for https://github.com/qutebrowser/qutebrowser/issues/1954. `https://www.google.com/search?q=for bar` is still supported by the remaining check (any `url.auto_search`). Yeah, it's a bit hacky.

**Pull request merged**